### ISSUE TITLE
Dynamic Padding in training script

### ISSUE THREAD

**Issue by ashish230897** [allenai/open-instruct#605]
Hello,

The static padding happening here in the code (open-instruct/open_instruct/grpo_vllm_thread_ray_gtrl.py):
```
g_response_token_ids = response_ids_Q.get()
                    DUMMY_PAD_TOKEN = (
                        args.stop_token_id
                    )  # we can't use tokenizer.pad_token_id because it's outside vocab and `torch.gather(all_logprob, 2, response.unsqueeze(-1))` will error out
g_padded_response_ids = [
    response + [DUMMY_PAD_TOKEN] * (args.response_length - len(response))
    for response in g_response_token_ids
]
g_padded_response_ids = torch.tensor(g_padded_response_ids, device=device)
g_vllm_responses[:] = g_padded_response_ids

dist.broadcast(g_vllm_responses, src=0)
local_vllm_responses = g_vllm_responses[
accelerator.process_index * queries.shape[0] : (accelerator.process_index + 1) * queries.shape[0]
]
```

can be changed to dynamic padding depending on the largest sequence length in the global batch:
```
g_response_token_ids = response_ids_Q.get()
                    DUMMY_PAD_TOKEN = args.stop_token_id  # we can't use tokenizer.pad_token_id because it's outside vocab and `torch.gather(all_logprob, 2, response.unsqueeze(-1))` will error out

# get max size of response ids in the batch
max_batch_size = torch.tensor(max(len(inner) for inner in g_response_token_ids), device=device, 
                            dtype=torch.long)
largest_size.copy_(max_batch_size).   # largest_size is defined above in the code similar to g_vllm_responses
g_padded_response_ids = [
    response + [DUMMY_PAD_TOKEN] * (args.response_length - len(response))
    for response in g_response_token_ids
]
g_padded_response_ids = torch.tensor(g_padded_response_ids, device=device)
g_vllm_responses[:] = g_padded_response_ids


dist.broadcast(g_vllm_responses, src=0)
dist.broadcast(largest_size, src=0)

local_vllm_responses = g_vllm_responses[
accelerator.process_index * queries.shape[0] : (accelerator.process_index + 1) * queries.shape[0]
]

local_vllm_responses = local_vllm_responses[:, :largest_size.item()]
```

this will help reduce the training time. Let me know if something looks wrong.

**Comment by ashish230897** [comment]
Got it, thanks!

**Comment by vwxyzjn** [comment]
Hey @ashish230897, feel free to check out https://github.com/allenai/open-instruct/blob/grpo-fast-pro/docs/algorithms/grpo.md#grpo_fastpy. It does the kind of thing you mentioned that it does not needlessly add pad tokens, but more generally using packing. 

Feel free to give it a try and let me know how it goes!

**Comment by ashish230897** [comment]
Hey @vwxyzjn, sure I will give it a try and let you know how it goes! Thanks.
